packages/swayze/lib/src/widgets/headers/gestures/header_gesture_detector.dart

```diff
@@ -250,67 +274,251 @@ class _HeaderGestureDetectorState extends State<HeaderGestureDetector> {
     );
   }
 
+  /// Handles drag starts that should start dragging a header around.
+  void handleStartDraggingHeader(
+    DragStartDetails details,
+    Range selectionRange,
+  ) {
+    /// Sets the dragging cursor to be basic.
+    /// Instead of using [SystemMouseCursors.grabbing], we set the basic cursors
+    /// because currently there is no mechanism to globally change the cursor on
+    /// desktop, this means that the cursor would be a closed hand only when
+    /// hovering the header, which would cause a weird change of cursors during
+    /// a drag action, making the user think that something went wrong with the
+    /// action.
+    setCursorState(SystemMouseCursors.basic);
+
+    Actions.invoke(
+      context,
+      HeaderDragStartIntent(
+        draggingPosition: details.localPosition,
+        headers: selectionRange,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  /// Handles header dragging updates.
+  void handleUpdateDraggingHeader(
+    DragUpdateDetails gestureDetails,
+    _HeaderGestureDetails details,
+  ) {
+    Actions.invoke(
+      context,
+      HeaderDragUpdateIntent(
+        draggingPosition: gestureDetails.localPosition,
+        header: details.headerPosition,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  void handleDragEnd(SwayzeHeaderDragState state) {
+    if (state.isDropAllowed) {
+      Actions.invoke(
+        context,
+        HeaderDragEndIntent(
+          header: state.dropAtIndex,
+          axis: widget.axis,
+        ),
+      );
+    } else {
+      handleDragCancel();
+    }
+  }
+
+  void handleDragCancel() {
+    Actions.invoke(
+      context,
+      HeaderDragCancelIntent(widget.axis),
+    );
+  }
+
+  /// Sets a new cursor state.
+  void setCursorState(MouseCursor newCursor) {
+    if (newCursor != cursor) {
+      setState(() => cursor = newCursor);
+    }
+  }
+
+  /// Checks if a header is selected.
+  bool isHeaderSelected(int position, Axis axis) =>
+      internalScope.controller.selection.isHeaderSelected(position, axis);
+
+  /// Finds the selection that is under position.
+  ///
+  /// Returns null if position header is not selected.
+  HeaderUserSelectionModel? hoverSelection(int position, Axis axis) {
+    final selectionController = internalScope.controller.selection;
+    final selections = selectionController.userSelectionState.selections;
+
+    for (final selection in selections) {
+      if (selection is HeaderUserSelectionModel && selection.axis == axis) {
+        final range = Range(selection.start, selection.end);
+        if (range.contains(position)) {
+          return selection;
+        }
+      }
+    }
+    return null;
+  }
+
+  /// Finds the range of the current reference selection and all adjacent
+  /// selections.
+  ///
+  /// Returns a range containing all adjacent selections from the reference
+  /// selection, so it can be dragged as a single group.
+  Range headerSelectionRange(HeaderUserSelectionModel referenceSelection) {
```

Comment:
Do we want to add tests for this? I think it would help make sure we don't break any assertions.

Reply:
yes, i'm thinking about how to get this out of the widget